File: android/app/src/main/java/com/sanity/urlcleaner/UrlCleaner.kt
package com.sanity.urlcleaner

import android.net.Uri
import java.util.regex.Pattern

object UrlCleaner {
    fun tryClean(text: String, rules: List<UrlRule>): String? {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return null
        if (!trimmed.startsWith("http://", ignoreCase = true) &&
            !trimmed.startsWith("https://", ignoreCase = true)) {
            return null
        }

        val uri = Uri.parse(trimmed)
        val scheme = uri.scheme?.lowercase()
        val host = uri.host
        if ((scheme != "http" && scheme != "https") || host.isNullOrEmpty()) return null

        var result = trimmed
        for (rule in rules) {
            if (!domainMatches(host, rule.domain)) continue
            try {
                result = Pattern.compile(rule.regex, Pattern.CASE_INSENSITIVE)
                    .matcher(result)
                    .replaceAll("")
            } catch (_: Exception) {
                // Skip invalid regex patterns in user config.
            }
        }

        result = tidyUrl(result)
        return if (result == trimmed) null else result
    }

    private fun domainMatches(host: String, domain: String): Boolean {
        if (domain.isEmpty() || domain == "*") return true
        val hostLower = host.lowercase()
        val domainLower = domain.lowercase()
        return hostLower == domainLower || hostLower.endsWith(".$domainLower")
    }

    private fun tidyUrl(url: String): String {
        var value = url
        val patterns = listOf("[?&]+$", "\\?&", "&&+")
        val replacements = listOf("", "?", "&")
        for ((pattern, replacement) in patterns.zip(replacements)) {
            value = Pattern.compile(pattern).matcher(value).replaceAll(replacement)
        }
        return value
    }
}

File: mac/Sanity/Sources/UrlCleaner.swift
import Foundation

enum UrlCleaner {
    static func tryClean(_ text: String, rules: [UrlRule]) -> String? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty,
              trimmed.lowercased().hasPrefix("http://") || trimmed.lowercased().hasPrefix("https://"),
              let url = URL(string: trimmed),
              let scheme = url.scheme?.lowercased(),
              scheme == "http" || scheme == "https",
              let host = url.host else {
            return nil
        }

        var result = trimmed
        for rule in rules where domainMatches(host: host, domain: rule.domain) {
            guard let regex = try? NSRegularExpression(pattern: rule.regex, options: [.caseInsensitive]) else {
                continue
            }
            let range = NSRange(result.startIndex..<result.endIndex, in: result)
            result = regex.stringByReplacingMatches(in: result, options: [], range: range, withTemplate: "")
        }

        result = tidyUrl(result)
        return result == trimmed ? nil : result
    }

    private static func domainMatches(host: String, domain: String) -> Bool {
        if domain.isEmpty || domain == "*" {
            return true
        }
        let hostLower = host.lowercased()
        let domainLower = domain.lowercased()
        return hostLower == domainLower || hostLower.hasSuffix("." + domainLower)
    }

    private static func tidyUrl(_ url: String) -> String {
        var value = url
        let patterns = [
            "[?&]+$",
            "\\?&",
            "&&+"
        ]
        let replacements = ["", "?", "&"]
        for (pattern, replacement) in zip(patterns, replacements) {
            guard let regex = try? NSRegularExpression(pattern: pattern) else { continue }
            let range = NSRange(value.startIndex..<value.endIndex, in: value)
            value = regex.stringByReplacingMatches(in: value, options: [], range: range, withTemplate: replacement)
        }
        return value
    }
}
